relentless/potential/pair.py

```diff
@@ -156,6 +157,213 @@ def _derivative(self, param, r, epsilon, sigma, **params):
+        rs = np.asarray(r, dtype=np.float64)
+        ks = np.asarray(u, dtype=np.float64)
+        if self.mode == 'diff':
+            # difference is next knot minus my knot, with last knot fixed at its current value
+            ks[:-1] -= ks[1:]
+
+        # make all knots variables, but hold all r and the last knot const
+        for i in range(self.num_knots):
+            self.coeff[pair]['r-{}'.format(i)] = core.Variable(rs[i],const=True)
+            self.coeff[pair]['knot-{}'.format(i)] = core.Variable(ks[i],const=(i==self.num_knots-1))
+
+    def _energy(self, pair, r, **params):
+        """Evaluates the spline potential energy.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        r : array_like
+            The value or values of r at which to evaluate the energy.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the energy evaluated at r.
+
+        """
+        r,u,s = self._zeros(r)
+        params = self.coeff[pair]
+        u = self._interpolate(params)(r)
+        if s:
+            u = u.item()
+        return u
+
+    def _force(self, pair, r, **params):
+        """Evaluates the spline force.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        r : array_like
+            The value or values of r at which to evaluate the force.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the force evaluated at r.
+
+        """
+        r,f,s = self._zeros(r)
+        params = self.coeff[pair]
+        f = -self._interpolate(params).derivative(r, 1)
+        if s:
+            f = f.item()
+        return f
+
+    def _derivative(self, pair, param, r, **params):
+        """Evaluates the spline parameter derivative using finite differencing.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to initialize the spline potential.
+        param : str
+            The parameter with respect to which to take the derivative (the knot coefficient).
+        r : array_like
+            The value or values of r at which to evaluate the derivative.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the derivative evaluated at r.
+
+        """
+        r,d,s = self._zeros(r)
+        h = 0.001
+        params = self.coeff[pair]
+
+        #perturb params[param]
+        params_high = copy.deepcopy(params.todict())
+        params_low = copy.deepcopy(params.todict())
+        params_high[param].value += h
+        params_low[param].value -= h
+
+        d = self._interpolate(params_high)(r) - self._interpolate(params_low)(r)
+        d /= (2*h)
+        if s:
+            d = d.item()
+        return d
```

Comment:
Anticipating that we might have other types of variables in a coefficient dictionary, do we want to force the user to leave the `Variable`s that we have created alone? We could override the parent `derivative` method to make sure everything stays that way:
```py
def derivative(self, pair, param, r):
    for i in range(self.num_knots):
        ri,ki = self._knot_params(i)
        if not isinstance(self.coeff[pair][ri], core.Variable):
            # raise error
        if not isinstance(self.coeff[pair][ki], core.Variable):
            # raise error
    return super().derivative(pair, param, r)
```
and then admonish them not to override them in the docs.